### ISSUE TITLE
Node: Add flush function

### ISSUE THREAD

**Issue by mikicho** [segmentio/analytics-next#822]
We need this for testing to empty the queue before assertion. 
`maxEventsInBatch` still do it async, which cause flaky tests

----
@segmentio update: team is starting the process of looking at a reimplemention of the vanilla flush function, though requirements are a bit more complex than the old node SDK because of the existence of async plugins. See: https://github.com/segmentio/analytics-next/issues/822#issuecomment-1776250023

**Comment by silesky** [comment]
Hey, this is the purpose of 'await analytics.closeAndFlush()'.

There is no way to do a synchronous flush.

**Comment by mikicho** [comment]
@silesky yeah, I don't want to close the connection, just flush all events in the queue.

> There is no way to do a synchronous flush.

I don't it to be sync, I can await for it:
```
analytics.flush()
```

**Comment by silesky** [comment]
@mikicho there is no persistent connection to close. The close means stopping analytics from accepting new events when it's in the process of closing. 

You can instantiate a new analytics after if you like afterward.

**Comment by mikicho** [comment]
@silesky why did you close this issue?
My case is valid, and `closeAndFlush` isn't good enough because it impacts the next test.

**Comment by silesky** [comment]
@mikicho sure: feel free to open a PR.

**Comment by mikicho** [comment]
> I do think in general that instantiating a server between tests is typically a good idea

Test cases or test suites? 😅 
I instantiate a server `beforeAll`.

> I am also not sure why you need a complex test suite just for our analytics library

I have something like that:

```
describe('Users', () => {
  beforeAll(() => {
    startServer() // this does almost production-like bootstrap
  })

  it('should fire "User Created" event', () => {
    const scope = nock('https://api.segment.io/v1/batch').reply(200); // https://github.com/nock/nock/

    await client.post('/users', {...});
    
   // await analytics.flush()
    expect(scope.isDone()).toBeTruthy()
  })
})
```

So I need a way to wait for the event that `POST /users` fires.

As a workaround, I copied `closeAndFlush` function and remove the `this._closed = true` but it's a little hacky and I access to `_pendingEvent` which should be private.

I just didn't think it will difficult to add a `flush` function but it seems like you oppose and I guess you have a good reason for that because the legacy packages do have a `flush` function and you decided to remove it.

Do you have other ideas on how to support my case?

**Comment by silesky** [comment]
I'm weary of introducing APIs that only get used for testing, as it increases the surface area and potential confusion, and kind of negates the idea of an e2e test reflecting 'reality'. 

Some options:
1. Reset the analytics module before each test - https://stackoverflow.com/questions/45571816/reset-single-module-with-jest would make sense.
2. Tear down the server before each analytics test and restart.
3. Do something hacky like (no warranty): 
```
 beforeEach(() => {
   analytics['_closed'] = false
 })
```

**Comment by silesky** [comment]
I'm weary of introducing APIs that only get used for testing, as it increases the surface area and potential confusion, and kind of negates the idea of an e2e test reflecting 'reality'. 

Try this:
``` ts
  it('should fire "User Created" event', () => {
   //  you want analytics to to be instantiated with maxEventsInBatch: 1 -- you can do this with mocks OR with a test env variable.
    const scope = nock('https://api.segment.io/v1/batch').reply(200); // https://github.com/nock/nock/
    const trackPromise = new Promise(resolve => analytics.once('track', resolve)) // assuming you sent a track event
    await client.post('/users', {...});
    await trackPromise
    expect(scope.isDone()).toBeTruthy()
  })
```

**Comment by silesky** [comment]
> There is a valid usecase for flush without close, and thats when you want to flush without close!
> For example, you may want to flush at the end of each request in lambda to ensure that events get dispatched before lambda suspends. Why would you want to reinitialize multiple times on a warm lambda?

1. The cost of reinitialization is non-existent; there is no reason to preserve state between lambda executions.
2. There are other ways to do this without an explicit flush function, as outlined in this thread.

**Comment by silesky** [comment]
Hey @makeable I edited my post above to remove the part about lambda execution, as that's not really relevant to the discussion. Sorry, I think you responded to the old comment.

**Comment by silesky** [comment]
Sorry, that sounds frustrating. We will be focusing on improving our documentation -- I see we don't have much emitter docs, this is a new repo/interface, and sometimes we don't really notice the gaps until people tell us.

I will also be discussing bring back an explicit flush method as more of a feature request -- we removed it because the main reason people were using it was for graceful shutdown, and it wasn't working as expected for obvious reasons. 

However, there is another subset of users who use it for testing / lamda stuff and the drained event is  a little confusing (especially if it's not a documented solution!).

**Comment by silesky** [comment]
@makeable oops, can't understand-> can understand

**Comment by silesky** [comment]
@makeable yep, but they are 'private' methods -- what is your comfort level there?

**Comment by chrisradek** [comment]
@makeable Thanks for engaging with us here!

Just to make sure we're all on the same page:
- You want to be able to instantiate your segment analytics class once outside a Lambda handler
- You need a way to know once all events from Segment are flushed, without instrumenting each API call

Is that right? Are you ok with calling `flush` on an analytics client within a handler wrapper?

> It seems segment is the only "service" that has issues keeping the event loop active, which is why we need to flush it at all.

For context, why do we even need flush at all? 2 reasons:
1. analytics does batching by default, so without the flush, your lambda function could be frozen before the batch of events is actually sent, and it may or may not be thawed based on your lambda load.
2. Events go through a pipeline of async plugins before getting batched. This allows you to do things like enrich events in a centralized location. This is really related to the first point - these events can be lost if the lambda function is frozen and never thawed before the HTTP request to segment is made.

Now, it would be great if we could make the HTTP request immediately if both the batch size is 1, and all plugins are synchronous, but that's a digression!

As @silesky mentioned, our guidance so far has been to instantiate a new analytics client within the Lambda handler (each lambda invocation), and call `closeAndFlush()` in your handler to ensure events make it to Segment. However, that definitely adds additional burden to you as you now need to make sure your analytics instance is passed through to any callsite that uses it, versus importing a singleton/relying on closures.

I think it makes sense to support a flush that doesn't close the client. One scenario we need to address is what to do with new events when there is a pending flush: wait to resolve flush until those have also been sent, or hold off on sending them until the current flush has completed. In your scenario, would you expect any track calls to be made after calling flush?

Sorry for the wall of text, thanks again for sharing your use-case with us!

**Comment by silesky** [comment]
@makeable @mikicho 

`analytics.flush` function was released in 1.3.0.

You can now do:
```ts
await analytics.flush()
```
